From 69675c09fd41cbc14c033d910d1920cd89ee3188 Mon Sep 17 00:00:00 2001
From: Nitin Shetty <quic_nitinsj@quicinc.com>
Date: Mon, 29 Nov 2021 16:25:37 +0530
Subject: [PATCH] net: pptp: set pptp interface in SKB

during pptp encap set incoming interface as pptp
in SKB.

Change-Id: Ia8a90285070f13f8a7cd94c6781905eea783d0cf
Signed-off-by: Nitin Shetty <quic_nitinsj@quicinc.com>
---
 drivers/net/ppp/pptp.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ppp/pptp.c b/drivers/net/ppp/pptp.c
index 660e211feb4a..c723535c8816 100644
--- a/drivers/net/ppp/pptp.c
+++ b/drivers/net/ppp/pptp.c
@@ -323,14 +323,18 @@ static int pptp_xmit(struct ppp_channel *chan, struct sk_buff *skb)
 
 	pptp_ifindex = ppp_dev_index(chan);
 
-	/* set incoming interface as the ppp interface */
-	if (skb->skb_iif)
-		skb->skb_iif = pptp_ifindex;
-
 	/* If the PPTP GRE seq number offload module is not enabled yet
 	 * then sends all PPTP GRE packets through linux network stack
 	 */
 	if (!opt->pptp_offload_mode) {
+		/* set incoming interface as the ppp interface */
+		pptp_dev = dev_get_by_index(&init_net, pptp_ifindex);
+		if (pptp_dev) {
+			skb->dev = pptp_dev;
+			skb->skb_iif = pptp_ifindex;
+			dev_put(pptp_dev);
+		}
+
 		ip_local_out(net, skb->sk, skb);
 		return 1;
 	}
-- 
2.34.1

